src/Uno.UI.Composition/Composition/Visual.skia.cs

```diff
@@ -178,21 +204,72 @@ internal virtual void Render(in DrawingSession parentSession)
 		}
 
 		using var session = BeginDrawing(in parentSession);
-		Draw(in session);
+
+		if (_requiresRepaint)
+		{
+			_requiresRepaint = false;
+			_recorder ??= new SKPictureRecorder();
+			var recorderSession = session with { Canvas = _recorder.BeginRecording(new SKRect(float.NegativeInfinity, float.NegativeInfinity, float.PositiveInfinity, float.PositiveInfinity)) };
+			// To debug what exactly gets repainted, replace the following line with `Draw(in session);`
+			Paint(in recorderSession);
+			_picture = _recorder.EndRecording();
+		}
+
+		if (_picture is { })
+		{
+			session.Canvas.DrawPicture(_picture);
+		}
+
+		// The CornerRadiusClip doesn't affect the visual itself, only its children
+		CornerRadiusClip?.Apply(session.Canvas, this);
+
+		foreach (var child in GetChildrenInRenderOrder())
+		{
+			child.Render(in session);
+		}
 	}
 
 	/// <summary>
 	/// Draws the content of this visual.
 	/// </summary>
 	/// <param name="session">The drawing session to use.</param>
-	internal virtual void Draw(in DrawingSession session)
+	internal virtual void Paint(in PaintingSession session) { }
+
+	private Vector3 GetTotalOffset()
```

Comment:
I feel like we may want to keep this `internal` **even if** it's only used in `Visual` class. In case someone wanted to access `Offset` externally, it may be useful that they see IntelliSense suggesting `GetTotalOffset()` so they have to think which one they need.

Reply:
I don't have a strong opinion on this, but I would argue that internal code that isn't inside Visual should _not_ need to read the "total offset" (or any offset really) but only set it. Keeping it private makes it clear that you should think twice if you really need to depend on it.